### DISCUSSION TITLE
Using hsmcpp in gamedev

### DISCUSSION THREAD

**Discussion by ershovdz** [igor-krechetov/hsmcpp#6] [Q&A] [answered]
I want to use hsmcpp in game render loop. Is it possible repeatedly execute some function for active states (including parent and substates) ?
May be I can add self transition to all states and send Update event something like that:

registerSelfTransition(MyState, UpdateEventID, TransitionType::INTERNAL_TRANSITION, []{ draw calls here })

In render loop:
hsm.transitionSync(UpdateEventID)

But if I understand right the documentation, self trasition will be executed only for the current active substate (not parent).

**Comment by igor-krechetov** [comment, marked as answer]
Hello @ershovdz. 

In general, the easiest way to repeatedly trigger some event for active states would be to use a repeating timer:
![self_transition_example](https://user-images.githubusercontent.com/58586005/232645316-2d30433d-71ea-4745-b3f2-79989f589e68.png)

_But if I understand right the documentation, self transition will be executed only for the current active substate (not parent)._
Yes, if both child and parent have exactly same event defined then only child's transition will be executed ([Priority of transitions](https://hsmcpp.readthedocs.io/en/latest/features/transitions/transitions.html#priority-of-transitions)). So no, it's impossible to execute some function for both parent and active child on a single event. But there are way to workaround it (for example by using 2 different events for parent and for children and triggering parent event from a child self-transition).

I have a feeling you are trying to do something like scene composition using state machine (so that each state only knows how to render part of the scene). If that's the case, this might not be the best approach. State machine can't guarantee execution time so you probably will end up with inconsistent frame rate or (if one of the callbacks would take too long to execute) even stalled animations.

I could recommend you to consider using [getActiveStates](https://hsmcpp.readthedocs.io/en/latest/api/classhsmcpp_1_1HierarchicalStateMachine.html#_CPPv4NK6hsmcpp24HierarchicalStateMachine15getActiveStatesEv) API inside your rendering loop to decide on the scene composition. This function returns a list of currently active states in accordance with HSM hierarchy. For example, if state **Child_1_2** is active call to this API will return [Parent, Child_1, Child_1_2]:

![get_active_states_example](https://user-images.githubusercontent.com/58586005/232655298-591a77c9-c51a-4a05-8d05-1d43959be687.png)

This way you could keep rendering and game rules/events separate and not block each other.

**Comment by ershovdz** [comment]
Keeping rendering and app state separate - seems to be good idea. I will try to use getActiveStates. Thanks for the tip !
